File: architecture-skills/design-patterns/examples/decorator/decorator.js
// Decorator pattern in JavaScript.
//
// Wraps an object with another object of the same shape. Not the same
// thing as the TC39 @decorator syntax, which rewrites class members at
// definition time; this pattern stacks wrappers around instances at
// runtime. For functions, higher-order functions play the same role.

class Espresso {
  description() {
    return "espresso";
  }
  cost() {
    return 2.0;
  }
}

class AddOn {
  constructor(inner, label, surcharge) {
    this.inner = inner;
    this.label = label;
    this.surcharge = surcharge;
  }
  description() {
    return `${this.inner.description()} + ${this.label}`;
  }
  cost() {
    return this.inner.cost() + this.surcharge;
  }
}

const milk = (inner) => new AddOn(inner, "milk", 0.3);
const mocha = (inner) => new AddOn(inner, "mocha", 0.5);
const whippedCream = (inner) => new AddOn(inner, "whipped cream", 0.4);

const plain = new Espresso();
const fancy = whippedCream(mocha(milk(new Espresso())));

for (const drink of [plain, fancy]) {
  console.log(`${drink.description()}: $${drink.cost().toFixed(2)}`);
}

File: architecture-skills/design-patterns/examples/decorator/decorator.py
"""Decorator pattern in Python.

Wraps an object with another object of the same interface to add
behavior at runtime. Related to but different from Python's @decorator
syntax, which wraps functions or classes once at definition time; this
pattern stacks wrappers around individual instances dynamically.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Beverage(ABC):
    @abstractmethod
    def description(self) -> str: ...

    @abstractmethod
    def cost(self) -> float: ...


class Espresso(Beverage):
    def description(self) -> str:
        return "espresso"

    def cost(self) -> float:
        return 2.00


class AddOn(Beverage):
    """Base decorator: holds the wrapped beverage."""

    def __init__(self, inner: Beverage) -> None:
        self._inner = inner


class Milk(AddOn):
    def description(self) -> str:
        return f"{self._inner.description()} + milk"

    def cost(self) -> float:
        return self._inner.cost() + 0.30


class Mocha(AddOn):
    def description(self) -> str:
        return f"{self._inner.description()} + mocha"

    def cost(self) -> float:
        return self._inner.cost() + 0.50


class WhippedCream(AddOn):
    def description(self) -> str:
        return f"{self._inner.description()} + whipped cream"

    def cost(self) -> float:
        return self._inner.cost() + 0.40


if __name__ == "__main__":
    plain = Espresso()
    fancy = WhippedCream(Mocha(Milk(Espresso())))
    for drink in (plain, fancy):
        print(f"{drink.description()}: ${drink.cost():.2f}")
